### PR TITLE
Code cleanup and compiler warn fixs

### PR DESCRIPTION
**Pull request by abderraouf-adjal**
I fixed a lot of compiler warnings when compile with the flags:

`-W -Wall -Wextra -Wformat=2 -Wstrict-prototypes -Wimplicit-function-declaration -Wredundant-decls -Wdeclaration-after-statement -Wconversion -Wsign-conversion -Winit-self -Wundef -Wshadow -Wpointer-arith -Wreturn-type -Wsign-compare -Wmultichar -Wformat-nonliteral -Wuninitialized -Wformat-security -pedantic`

Also there is a warning when compile with `-O1` or `-O2`, I did not try to fix it.

### CODE REVIEW

File: raqm.c [hunk 1]
```diff
@@ -457,30 +461,34 @@ raqm_shape_u32 (unsigned int* u32_str,
...
-    FriBidiCharType* types;
-    FriBidiLevel* levels;
-    FriBidiParType par_type;
-    hb_script_t* scripts;
-    hb_unicode_funcs_t* unicode_funcs;
-    hb_script_t last_script_value;
-    int last_script_index = -1;
-    int last_set_index = -1;
-    Stack* script_stack;
+    int j = 0;
```
Comment (khaledhosny) [inline raqm.c]: This variable is only used in the for loop below, so it is better be declared there. You just need to move it to the top of the block to fix the compiler warning.

File: Makefile [hunk 2]
```diff
@@ -1,6 +1,6 @@
 PAKAGES = freetype2 harfbuzz fribidi
 
-CFLAGS = `pkg-config --cflags $(PAKAGES)` -DTESTING -Wall -ansi 
+CFLAGS = `pkg-config --cflags $(PAKAGES)` -DTESTING -std=c99 -Wall
```
Comment (khaledhosny) [inline Makefile]: It would be nice to have all the warnings you were compiling with enable here, to avoid there recurrence in the future.

File: raqm.c [hunk 3]
```diff
@@ -35,26 +35,24 @@
 typedef struct _FriBidiRunStruct FriBidiRun;
 struct _FriBidiRunStruct
 {
-  FriBidiRun *prev;
-  FriBidiRun *next;
+    FriBidiRun *prev;
+    FriBidiRun *next;
```
Comment (khaledhosny) [inline raqm.c]: I’d avoid the indentation and other stylistic changes in this code, it follows FriBiDi code style because it was written to be submitted to FriBiDi. Though I think it is OK to change it to follow the style of the rest of the code, but in a separate commit please.

File: raqm.c [hunk 4]
```diff
@@ -196,8 +203,8 @@ stack_create (int max)
 {
     Stack* stack;
     stack = (Stack*) malloc (sizeof (Stack));
-    stack->scripts = (hb_script_t*) malloc (sizeof (hb_script_t) * max);
-    stack->pair_index = (int*) malloc (sizeof (int) * max);
+    stack->scripts = (hb_script_t*) malloc (sizeof (hb_script_t) * (size_t)(max));
+    stack->pair_index = (int*) malloc (sizeof (int) * (size_t)(max));
```
Comment (khaledhosny) [inline raqm.c]: We can just make `max` (and `capacity`) of type `size_t` and avoid the cast altogether.

File: raqm.c [hunk 5]
```diff
@@ -285,8 +290,7 @@ itemize_by_script(int bidirun_count,
                  FriBidiRun *bidiruns,
                  Run *runs)
 {
-    int i, j;
-    int run_count = 0;
+    int i, j, run_count = 0;
```
Comment (khaledhosny) [inline raqm.c]: This change is not really needed in this commit, mixing unrelated changes in the same commit makes it harder to review the code.

File: raqm.c [hunk 6]
```diff
@@ -380,7 +384,7 @@ harfbuzz_shape (FriBidiChar* unicode_str,
     run->hb_buffer = hb_buffer_create ();
 
     /* adding text to current buffer */
-    hb_buffer_add_utf32 (run->hb_buffer, unicode_str, length, run->pos, run->len);
+    hb_buffer_add_utf32 (run->hb_buffer, unicode_str, length, (unsigned int)(run->pos), run->len);
```
Comment (khaledhosny) [inline raqm.c]: We can declare `pos` as an `nsigned int` to avoid the cast, it is only there for this function.

File: raqm.c [hunk 7]
```diff
@@ -573,7 +581,6 @@ raqm_shape_u32 (unsigned int* u32_str,
         }
         else
         {
-            int j;
```
Comment (khaledhosny) [inline raqm.c]: Why was this declaration moved up, it is better to limit the scope of the variables.

File: raqm.c [hunk 8]
```diff
@@ -649,14 +656,14 @@ raqm_shape_u32 (unsigned int* u32_str,
     {
         hb_glyph_info = hb_buffer_get_glyph_infos (runs[i].hb_buffer, &glyph_count);
         hb_glyph_position = hb_buffer_get_glyph_positions (runs[i].hb_buffer, &postion_length);
-        int j;
-        for (j = 0; j < glyph_count; j++)
+
+        for (k = 0; k < glyph_count; k++)
```
Comment (khaledhosny) [inline raqm.c]: Why `k` and not `j`.

File: raqm.c [hunk 9]
```diff
@@ -649,14 +656,14 @@ raqm_shape_u32 (unsigned int* u32_str,
...
         {
-            glyph_info[index].index = hb_glyph_info[j].codepoint;
-            glyph_info[index].x_offset = hb_glyph_position[j].x_offset;
-            glyph_info[index].y_offset = hb_glyph_position[j].y_offset;
-            glyph_info[index].x_advance = hb_glyph_position[j].x_advance;
-            glyph_info[index].cluster = hb_glyph_info[j].cluster;
+            glyph_info[index].index = (int)(hb_glyph_info[k].codepoint);
```
Comment (khaledhosny) [inline raqm.c]: I think `index` should be declared as `unsigned int` which would avoid the need of the cast (and is the correct thing to do).

File: raqm.c (hunk 4 above)
Comment (abderraouf-adjal) [inline raqm.c]: > We can just make max (and capacity) of type size_t and avoid the cast altogether.

I was going to write an issue about the types that are used in the lib.

File: raqm.c (hunk 7 above)
Comment (abderraouf-adjal) [inline raqm.c]: > Why was this declaration moved up, it is better to limit the scope of the variables.

Probably because of tiny wan about declaration in the middle of function code.

File: raqm.c (hunk 8 above)
Comment (abderraouf-adjal) [inline raqm.c]: > Why k and not j.

To use an `unsigned` type.

**Comment by abderraouf-adjal** [comment]
@khaledhosny changing variables sign in functions and `struct`'s need a different pull.

I need a summary of the most *important* changes this pull need if it will be merged.

**Comment by khaledhosny** [comment]
If it is not much work, I’d break it into one commit per warning, and all stylistic changes in one commit. Fixing compiler warnings is something I consider important, so thank you very much for taking care of this.

File: raqm.c (hunk 4 above)
Comment (khaledhosny) [inline raqm.c]: No great though have been put into choosing the types, the code just evolved “organically” as we tried things out.

File: raqm.c (hunk 7 above)
Comment (khaledhosny) [inline raqm.c]: Hmm, I though a declaration on top of a code block is valid in C89, what was the exact warning?

File: raqm.c (hunk 8 above)
Comment (khaledhosny) [inline raqm.c]: I see, I’d just move `j` declaration 2 lines above and make it `unsigned` as well.

File: raqm.c (hunk 7 above)
Comment (abderraouf-adjal) [inline raqm.c]: The change was to make the declarations `raqm.c:652` and `raqm.c:576` one declaration.

File: raqm.c (hunk 8 above)
Comment (abderraouf-adjal) [inline raqm.c]: `j` I did that in the 1st try, but the result is other warnings.

**Comment by khaledhosny** [comment]
If you can drop the unrelated stylistic changes, I’ll be happy to merge this and handle the other issues later.

File: raqm.c (hunk 7 above)
Comment (khaledhosny) [inline raqm.c]: I prefer limiting the scope of the variable as much as possible.

**Comment by abderraouf-adjal** [comment]
@khaledhosny I did a fast change, If there is tiny necessary changes point it :)

File: raqm.c [hunk 10]
```diff
@@ -51,11 +51,11 @@ static void reverse_run (FriBidiRun *arr, const FriBidiStrIndex len)
...
-      FriBidiRun temp = arr[i];
-      arr[i] = arr[len - 1 - i];
-      arr[len - 1 - i] = temp;
-    }
+  {
+    FriBidiRun temp = arr[i];
+    arr[i] = arr[len - 1 - i];
+    arr[len - 1 - i] = temp;
+  }
 }
```
Comment (khaledhosny) [inline raqm.c]: There are still stylistic changes here.

**Comment by khaledhosny** [comment]
OK, can you squash the two commits together?

File: raqm.c [hunk 11]
```diff
@@ -70,80 +70,89 @@ FriBidiStrIndex fribidi_reorder_runs (
...
+  FriBidiStrIndex i;
   FriBidiStrIndex run_count = 0;
   FriBidiStrIndex run_start = 0;
   FriBidiStrIndex run_index = 0;
 
- if
-    (len == 0)
-    {
-      goto out;
-    }
```
Comment (khaledhosny) [inline raqm.c]: Stylistic changes here as well.

**Comment by khaledhosny** [comment]
Merged manually as 18543862e6f34aec07608173fcca499d4b9f8b9a and eb127d247e5d7a01a05455a53022bbafa7756559.
